stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -208,6 +216,103 @@ public void run() {
+    return mDOMProvider.postAndWait(new UncheckedCallable<PerformSearchResponse>() {
+      @Override
+      public PerformSearchResponse call() {
+        final PerformSearchRequest request = mObjectMapper.convertValue(
+            params,
+            PerformSearchRequest.class);
+
+        // Searches are case insensitive so convert
+        // the query to lower case for simpler comparisons.
+        final String lowerCaseQuery = request.query.toLowerCase();
+
+        final Object root = mShadowDOM.getRootElement();
+
+        // Each search action has a unique ID so that
+        // it can be queried later.
+        final String id = UUID.randomUUID().toString();
+
+        final ArrayList<Integer> results = new ArrayList<>();
+
+        findMatches(lowerCaseQuery, root, results);
+
+        mSearchResults.put(id, results);
+
+        final PerformSearchResponse response = new PerformSearchResponse();
+        response.searchId = id;
+        response.resultCount = results.size();
+
+        return response;
+      }
+    });
+  }
+
+  private void findMatches(String query, Object element, List<Integer> resultIds) {
+    final ElementInfo info = mShadowDOM.getElementInfo(element);
+
+    for (int i = 0, size = info.children.size(); i < size; i++) {
+      final Object childElement = info.children.get(i);
+
+      if (doesElementMatch(query, childElement)) {
+        resultIds.add(mObjectIdMapper.getIdForObject(childElement));
+      }
+
+      findMatches(query, childElement, resultIds);
+    }
+  }
+
+  private boolean doesElementMatch(final String query, final Object element) {
+      final NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
+
+      final AttributeListAccumulator accumulator = getCachedAttributeListAccumulator();
+
+      try {
+        descriptor.getAttributes(element, accumulator);
+
+        for (int i = 0, size = accumulator.size(); i < size; i++) {
+          if (accumulator.get(i).toLowerCase().contains(query)) {
+            return true;
+          }
+        }
+      } finally {
+        returnCachedAttributeListAccumulator(accumulator);
+      }
+
+      return descriptor.getNodeName(element).toLowerCase().contains(query);
+  }
+
+  @ChromeDevtoolsMethod
+  public GetSearchResultsResponse getSearchResults(JsonRpcPeer peer, JSONObject params) {
+    final GetSearchResultsRequest request = mObjectMapper.convertValue(
+        params,
+        GetSearchResultsRequest.class);
+
+    final List<Integer> results = mSearchResults.get(request.searchId);
+
+    if (results == null) {
+      return null;
+    }
+
+    final List<Integer> resultsRange = results.subList(request.fromIndex, request.toIndex);
+
+    final GetSearchResultsResponse response = new GetSearchResultsResponse();
+    response.nodeIds = resultsRange;
+
+    return response;
+  }
+
+  @ChromeDevtoolsMethod
+  public void discardSearchResults(JsonRpcPeer peer, JSONObject params) {
```

Comment:
Ok, fair enough, looks like your usage of Map and this implementation is the right strategy then.